### DISCUSSION TITLE
Trigger workflow on Record creation

### DISCUSSION THREAD

**Discussion by rochmirabeau** [YetiForceCompany/YetiForceCRM#15431] [How to] [answered]
I have a module with a custom function that calculates a few totals.  I've set it to trigger on Edit/Creation; however, it seems to be only triggering on edit.  How can I make it trigger when I first create the record?

EDIT: public $executeImmediately = true;

**Comment by mariuszkrzaczkowski** [comment]
show all workflow steps

**Reply by rochmirabeau** [reply]
I wrote a simpler function for examples sake.  I'd like for this to execute on creation and on editing:

    public static function SetArea(Vtiger_Record_Model $recordModel) { 
        $length_inches = $recordModel->get('length_feet') * 12 + $recordModel->get('length_inches');
        $width_inches = $recordModel->get('width_feet') * 12 + $recordModel->get('width_inches');
        $area_sqft = ($length_inches * $width_inches) / 144; 
        $recordModel->set('area_sqft', $area_sqft);
        $recordModel->isNew = false;
        $recordModel->setHandlerExceptions(['disableHandlerClasses' => ['Vtiger_Workflow_Handler']]);
        $recordModel->save(); 
    }


I think it might have something to do with #13617

I could not get the conditional for isNew to work.  isNew = false works on edit only, isNew = true duplicates entry with one empty and one correct.

**Comment by vovpff** [comment]
@mariuszkrzaczkowski I think I has the same bug. I has selfmade function what send email from template. If `isNew() === 1` function should send one template, if `!== 1` than other one template. This is not working in my case for new records. If record edited than all perfect

```
    		$recordNew = $recordModel->isNew();
...
 if ($recordNew === 1) {
                    \App\Mailer::sendFromTemplate(['template' => 'LeadNew', 'moduleName' => 'Leads', 'recordId' => $recordModel->getId() , 'to' => $mails,
                                                   // 'cc' => $mailsCC,
                                                    'bcc' => $harcodedBcc, 
                                                  ]);
     } else {
                     \App\Mailer::sendFromTemplate(['template' => 'LeadNewwww', 'moduleName' => 'Leads', 'recordId' => $recordModel->getId() , 'to' => $mails,
                                                   ]);
     }
```

In my case main `if` condition not working

**Reply by rskrzypczak** [reply]
> In my case main if condition not working

Maybe because the isNew() function returns a bool, not a numeric value.

```php
 if ($recordNew) {
          \App\Mailer::sendFromTemplate(['template' => 'LeadNew', 'moduleName' => 'Leads', 'recordId' => $recordModel->getId() , 'to' => $mails,   'bcc' => $harcodedBcc,    ]);
  } else {
          \App\Mailer::sendFromTemplate(['template' => 'LeadNewwww', 'moduleName' => 'Leads', 'recordId' => $recordModel->getId() , 'to' => $mails,                                                  ]);
  }

**Reply by vovpff** [reply]
@rskrzypczak I will check

**Comment by mariuszkrzaczkowski** [comment]
it depends on the configuration of the first workflow step

**Reply by rochmirabeau** [reply]
Pardon my confusion, where should I be looking to configure the first step?

**Reply by mariuszkrzaczkowski** [reply]
1 Schedule workflow
https://gitdeveloper.yetiforce.com/index.php?module=Workflows&parent=Settings&view=Edit

**Comment by rochmirabeau** [comment, marked as answer]
What ended up working for me was the following conditional, previously explained in #13617 by @davide-alghi :

    public function workflowOnCreateAndEdit($copyRecordModel) {
    if ($copyRecordModel->isNew) {
        $recordModel = \Vtiger_Record_Model::getCleanInstance($copyRecordModel->getModuleName());
        $recordModel->setData($copyRecordModel->getData());
                $recordModel->ext = $copyRecordModel->ext;
                $recordModel->isNew = false;
            } else {
                $recordModel = $copyRecordModel;
            }
    }
    /* Insert your edits */
    $recordModel->setHandlerExceptions(['disableHandlerClasses' => ['Vtiger_Workflow_Handler']]);
    $recordModel->save(); 
    }

This works on create and edit.
